### DISCUSSION TITLE
[8.x] Primitive castes are not handled

### DISCUSSION THREAD

**Discussion by andrey-helldar** [laravel/framework#34387] [General]
- Laravel Version: 8.4.0
- PHP Version: 7.4.10
- Database Driver & Version: Postgres 12.4

### Description:
I am transferring an application from a MySQL 8 database to PostgreSQL 12.4 and I get this error:

```
SQLSTATE[22P02]: Invalid text representation: 7 ERROR:  invalid input syntax for type integer: "3.000"
(SQL: insert into "goods" ("id", "title", "good_type_id", "good_brand_id", "good_group_id", "params", "deleted_at", "quantity", "updated_at", "created_at") values (2686, 126 FANFARE LUXIO, 3, 9, 360, {"color":"#1368f5"}, ?, 3.000, 2020-09-17 13:44:55, 2020-09-17 13:44:55))
```

// Model casts:
![image](https://user-images.githubusercontent.com/10347617/93467078-ea4fdd80-f8f5-11ea-87a3-0d3595526910.png)

I checked the data before inserting and I see that the field did not pass the cast:
![image](https://user-images.githubusercontent.com/10347617/93467024-d3a98680-f8f5-11ea-8bbd-93a7e233bc5a.png)

The problem is that primitive variables are not cast before being written to the database:
https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L636-L672

There are two solutions:

1. Add method:
```php
protected function setQuantityAttribute($value)
{
    $this->attributes['quantity'] = (int) $value;
}
```

2. Add custom cast:
```php
<?php

namespace App\Casts\Primitives;

use Illuminate\Contracts\Database\Eloquent\CastsAttributes;

class IntegerCast implements CastsAttributes
{
    public function get($model, $key, $value, $attributes)
    {
        return $value;
    }

    public function set($model, $key, $value, $attributes)
    {
        return (int) $value;
    }
}
```

I consider the first as an overhead on primitive types, and generally keep quiet about the second.

**Comment by mfn** [comment]
Indeed, seems you're a victim of Laravel never performed casts on primitive types for setting them (only reading) and MySQL non-strict behaviour silently accepting invalid values.

I made the move from MySQL to PgSql years ago; interestingly I never noticed the absence of this; maybe I just got lucky. OTOH I'm also not developing project without static anlayzers and ide-helper, so nowadays I can't make that mistake even with the actual behviour missing.

Curious. Wonder what's the reason for this.

**Comment by andrey-helldar** [comment]
A value is given to the input, validated according to the integer rule - the validator passes it, then everything is correct. MySQL casts itself to the desired type, but in the case of Postgres this did not happen. I started looking for the reason and saw the lack of processing of primitive types. I had to create a custom cast and assign all integer fields in the project.
